sys/linux/socket_ip_tunnel.txt

```diff
@@ -0,0 +1,75 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <uapi/linux/if_tunnel.h>
+include <uapi/linux/ip6_tunnel.h>
+include <uapi/linux/if.h>
+include <uapi/linux/in.h>
+include <uapi/linux/ip.h>
+include <net/ip.h>
+
+# Fallback tunnels and placeholder names "syztnl*".
+ipv4_tunnel_names = "ip_vti0", "tunl0", "gre0", "gretap0", "erspan0", "syztnl0", "syztnl1", "syztnl2"
+ipv6_tunnel_names = "ip6tnl0", "ip6_vti0", "ip6gre0", "sit0", "syztnl0", "syztnl1", "syztnl2"
+
+ip_tunnel_protocols = 0, IPPROTO_IPIP, IPPROTO_GRE, IPPROTO_IPV6
+ip_tunnel_io_flags = GRE_CSUM, GRE_ROUTING, GRE_KEY, GRE_SEQ, GRE_STRICT, GRE_REC, GRE_ACK, GRE_FLAGS, GRE_VERSION, VTI_ISVTI
+
+type ipv4_header_only ipv4_packet_t[flags[ip_tunnel_protocols, int8], void]
+
+# IPv4 tunnel parameters
+ip_tunnel_parm {
+	name	string[ipv4_tunnel_names, IFNAMSIZ]
+	link	int32
+	i_flags	flags[ip_tunnel_io_flags, int16be]
+	o_flags	flags[ip_tunnel_io_flags, int16be]
+	i_key	int32be
+	o_key	int32be
+	iph	ipv4_header_only
+}
+
+ipv6_tunnel_flags = IP6_TNL_F_IGN_ENCAP_LIMIT, IP6_TNL_F_USE_ORIG_TCLASS, IP6_TNL_F_USE_ORIG_FLOWLABEL, IP6_TNL_F_MIP6_DEV, IP6_TNL_F_RCV_DSCP_COPY, IP6_TNL_F_USE_ORIG_FWMARK, IP6_TNL_F_ALLOW_LOCAL_REMOTE
+
+# IPv6 tunnel parameters
+ip6_tnl_parm2 {
+	name		string[ipv6_tunnel_names, IFNAMSIZ]
+	link		int32
+	proto		int8
+	encap_limit	int8
+	hop_limit	int8
+	flowinfo	int32be
+	flags		flags[ipv6_tunnel_flags, int32]
+	laddr		ipv6_addr
+	raddr		ipv6_addr
+	i_flags		flags[ip_tunnel_io_flags, int16be]
+	o_flags		flags[ip_tunnel_io_flags, int16be]
+	i_key		int32be
+	o_key		int32be
+}
+
+type ifreq_ipv4_tunnel ifreq_dev_t[ipv4_tunnel_names, ptr[inout, ip_tunnel_parm]]
+type ifreq_ipv6_tunnel ifreq_dev_t[ipv6_tunnel_names, ptr[inout, ip6_tnl_parm2]]
+
+# Repeat IP tunnel syscall variants here until #1913 is resolved.
+# IPv4 ioctls
+ioctl$sock_ipv4_tunnel_SIOCGETTUNNEL(fd sock_udp, cmd const[SIOCGETTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+ioctl$sock_ipv4_tunnel_SIOCADDTUNNEL(fd sock_udp, cmd const[SIOCADDTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+ioctl$sock_ipv4_tunnel_SIOCDELTUNNEL(fd sock_udp, cmd const[SIOCDELTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+ioctl$sock_ipv4_tunnel_SIOCCHGTUNNEL(fd sock_udp, cmd const[SIOCCHGTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+
+# IPv6 ioctls
+# SIOC***TUNNEL
+ioctl$sock_ipv6_tunnel_SIOCGETTUNNEL(fd sock_udp6, cmd const[SIOCGETTUNNEL], arg ptr[inout, ifreq_ipv6_tunnel])
```

Comment:
I see that `ip6gre_tunnel_ioctl()` and `vti6_ioctl()` are expecting `ip6_tnl_parm2` as an argument, while `ip6_tnl_ioctl()` expects `ip6_tnl_parm`. Maybe we need differentiation between those types of tunnels?

Reply:
1. Tracking the types of IP tunnels are hard. They are identified by name, so after we create a new tunnel with a specific type, we need to keep track of strings (e.g., what type of `syztnl0` is). I don't think it is possible to do that in syzkaller now?
2. [`ip6_tnl_parm2`](https://elixir.bootlin.com/linux/v5.8-rc3/source/include/uapi/linux/ip6_tunnel.h#L39) are exactly [`ip6_tnl_parm`](https://elixir.bootlin.com/linux/v5.8-rc3/source/include/uapi/linux/ip6_tunnel.h#L27) with some field added at the end. Thus, I assume that if we pass `ip6_tnl_parm2` into function that is expecting a pointer of `ip6_tnl_parm`, it will be fine.
3. [iproute2](https://github.com/shemminger/iproute2/search?q=ip6_tnl_parm&unscoped_q=ip6_tnl_parm) does not use `ip6_tnl_parm`.
Thus, I think stick with only `ip6_tnl_parm2`. I would like to hear your thoughts. Thanks!

Reply:
OK, seems reasonable, let's keep this as is.